File: recursive.go
package main

func getGroups(grid [][]bool) [][][]int {
	visited := make([][]bool, len(grid))
	for i := range visited {
		visited[i] = make([]bool, len(grid))
	}
	groups := [][][]int{}
	for column := 0; column < len(grid); column++ {
		for row := 0; row < len(grid[0]); row++ {
			group := getGroup(row, column, grid, visited, [][]int{})
			if len(group) > 0 {
				groups = append(groups, group)
			}
		}
	}
	return groups
}

func getGroup(row, column int, grid [][]bool, visited [][]bool, group [][]int) [][]int {
	if row < 0 || column < 0 || row >= len(grid[0]) || column >= len(grid) {
		return group
	}
	if !grid[column][row] || visited[column][row] {
		return group
	}
	visited[column][row] = true
	group = append(group, []int{column, row})
	group = getGroup(row-1, column-1, grid, visited, group)
	group = getGroup(row, column-1, grid, visited, group)
	group = getGroup(row+1, column-1, grid, visited, group)
	group = getGroup(row-1, column, grid, visited, group)
	group = getGroup(row+1, column, grid, visited, group)
	group = getGroup(row-1, column+1, grid, visited, group)
	group = getGroup(row, column+1, grid, visited, group)
	group = getGroup(row+1, column+1, grid, visited, group)
	return group
}

File: recursive.js
function getGroups(grid) {
  const visited = [...Array(grid.length)].map(row => Array(grid[0].length).fill(false));
  const groups = [];
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      const group = getGroup(i, j, grid, visited, []);
      if (group.length) {
        groups.push(group);
      }
    }
  }
  return groups;
}

function getGroup(i, j, grid, visited, group) {
  if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length) {
    return group;
  }
  if (!grid[i][j] || visited[i][j]) {
    return group;
  }
  visited[i][j] = true;
  group.push([i, j]);
  getGroup(i-1, j-1, grid, visited, group);
  getGroup(i-1, j,   grid, visited, group);
  getGroup(i-1, j+1, grid, visited, group);
  getGroup(i,   j-1, grid, visited, group);
  getGroup(i,   j+1, grid, visited, group);
  getGroup(i+1, j-1, grid, visited, group);
  getGroup(i+1, j,   grid, visited, group);
  getGroup(i+1, j+1, grid, visited, group);
  return group;
}

console.log(getGroups([
  [true, false, false, true],
  [false, false, true, true],
  [true, true, false, false],
  [true, false, false, false]
]));